pkg/network/transport/transport.go

```diff
@@ -0,0 +1,306 @@
+	})
+	if err != nil {
+		return fmt.Errorf("%w: %v", ErrListenerFailed, err)
+	}
+
+	t.ctx, t.cancel = context.WithCancel(context.Background())
+	t.listener = listener
+	t.acceptLoopWg.Add(1)
+	go func() {
+		defer t.acceptLoopWg.Done()
+		t.acceptLoop()
+	}()
+	return nil
+}
+
+// Stop gracefully shuts down the transport and all active connections.
+// Waits for the accept loop to finish before returning.
+func (t *Transport) Stop() error {
+	t.cancel()
+	if t.listener != nil {
+		t.listener.Close()
+	}
+	t.acceptLoopWg.Wait() // Wait for acceptLoop to finish
+	return nil
+}
+
+// Connect initiates a connection to a remote peer.
+// Returns the new connection or an error if connection fails.
+func (t *Transport) Connect(addr string) (*Conn, error) {
+	tlsConf := &tls.Config{
+		Certificates:       []tls.Certificate{*t.config.TLSCert},
+		NextProtos:         t.config.Handler.GetProtocols(),
+		ClientAuth:         tls.RequireAnyClientCert,
+		MinVersion:         tls.VersionTLS13,
+		InsecureSkipVerify: true,
+		VerifyPeerCertificate: func(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+			c, err := x509.ParseCertificate(rawCerts[0])
+			if err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			if err := t.config.CertValidator.ValidateCertificate(c); err != nil {
+				return fmt.Errorf("%w: %v", ErrInvalidCertificate, err)
+			}
+			return nil
+		},
+	}
+
+	quicConn, err := quic.DialAddr(t.ctx, addr, tlsConf, &quic.Config{
+		EnableDatagrams: true,
+		MaxIdleTimeout:  MaxIdleTimeout,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("%w: %v", ErrDialFailed, err)
+	}
+
+	conn := t.handleConnection(quicConn)
+	if conn == nil {
+		return nil, ErrConnFailed
+	}
+	return conn, nil
+}
+
+// GetConnection retrieves an active connection by peer key.
+// Returns the connection and whether it was found.
+func (t *Transport) GetConnection(peerKey string) (*Conn, bool) {
+	t.mu.RLock()
+	conn, ok := t.conns[peerKey]
+	t.mu.RUnlock()
+	return conn, ok
+}
+
+// ListConnections returns a slice of all active connections.
+func (t *Transport) ListConnections() []*Conn {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	conns := make([]*Conn, 0, len(t.conns))
+	for _, conn := range t.conns {
+		conns = append(conns, conn)
+	}
+	return conns
+}
+
+// acceptLoop continuously accepts incoming connections
+func (t *Transport) acceptLoop() {
+	defer t.cancel()
+	for {
+		select {
+		case <-t.ctx.Done():
+			return
+		default:
+			conn, err := t.listener.Accept(t.ctx)
+			if err != nil {
+				// Only log if it's not due to context cancellation/listener closing
+				if t.ctx.Err() == nil {
+					fmt.Printf("Failed to accept connection: %v\n", err)
+				}
+				if t.ctx.Err() != nil {
+					return
+				}
+				continue
+			}
+
+			go t.handleConnection(conn)
+		}
+	}
+}
+
+// handleConnection processes a new QUIC connection
+func (t *Transport) handleConnection(qConn quic.Connection) *Conn {
+	tlsState := qConn.ConnectionState().TLS
```

Comment:
Might be good to note that this is an unstable API (has a warning on the struct), if we upgrade later it could break.

Reply:
Actually couldn't we just do the connection validation in the `VerifyPeerCertifcate` callback too? Since you only need the negotiated protocol.

Reply:
Actually seems you can also use the `VerifyConnection` callback on `tls.Config` to do this.